### ISSUE TITLE
`needless_raw_strings`, `needless_raw_string_hashes`: False negative with macros

### ISSUE THREAD

**Issue by musjj** [rust-lang/rust-clippy#13503]
### Summary

[`needless_raw_strings`](https://rust-lang.github.io/rust-clippy/master/index.html#needless_raw_strings) and [`needless_raw_string_hashes`](https://rust-lang.github.io/rust-clippy/master/index.html#needless_raw_string_hashes) does not work when the string is inside macros.

### Lint Name

needless_raw_strings, needless_raw_string_hashes

### Reproducer

```rust
#![warn(clippy::needless_raw_strings, clippy::needless_raw_string_hashes)]

let _ = r"SELECT * FROM posts"; // needless-raw-strings
let _ = r#"SELECT * FROM posts"#; // needless-raw-strings
let _ = r##"SELECT * FROM "posts""##; // needless-raw-string-hashes

// None of lines below triggers any warning
println!(r"SELECT * FROM posts");
println!(r#"SELECT * FROM posts"#);
println!(r##"SELECT * FROM "posts""##);

// No warning when used with sqlx
sqlx::query!(
    r#"
    SELECT * FROM posts
    "#,
)
.fetch_all(&pool)
.await
.unwrap();
```

### Version

rustc 1.83.0-nightly (363ae4188 2024-09-24)
binary: rustc
commit-hash: 363ae4188316b8b22cf6c1890bc73d84d05f70a4
commit-date: 2024-09-24
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0

**Comment by samueltardieu** [comment]
The problem is not with macros, it is with `format_args!()` templates (and with macros using them, such as `println!()`).

**Comment by musjj** [comment]
It looks like that the issue is fixed in `clippy 0.1.84 (c1db4dc242 2024-10-25)` for `println!`. But it's still a problem for the [`sqlx`](https://github.com/launchbadge/sqlx) macro:

```rust
sqlx::query!(r#"SELECT * FROM posts"#) // no warnings
    .fetch_all(&pool)
    .await?;
```

The macro is implemented here: https://github.com/launchbadge/sqlx/blob/028084bce3a741e995c3e6c559c6dbb27a62534d/src/macros/mod.rs#L314. Is there something that is confusing clippy here?

**Comment by samueltardieu** [comment]
In general, Clippy doesn't lint the output of macros if it cannot determine that the tokens come from the source. Here, the procedural macro in sqlx seems to cut the link between the string literal and the emitted string. Clippy never sees the original string as it has been replaced by the macro expansion.
